internal/storage/fs/snapshot.go

```diff
@@ -1,14 +1,616 @@
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	indexFile = ".flipt.yml"
+	defaultNs = "default"
 )
 
-type storeSnapshot struct{}
+var ErrNotImplemented = errors.New("not implemented")
+
+// FliptIndex represents the structure of a well-known file ".flipt.yml"
+// at the root of an FS.
+type FliptIndex struct {
+	Version    string   `yaml:"version,omitempty"`
+	Inclusions []string `yaml:"inclusions,omitempty"`
+	Exclusions []string `yaml:"exclusions,omitempty"`
+}
+
+// storeSnapshot contains the structures necessary for serving
+// flag state to a client.
+type storeSnapshot struct {
+	ns        map[string]*namespace
+	evalDists map[string][]*storage.EvaluationDistribution
+}
+
+type namespace struct {
+	resource  *flipt.Namespace
+	flags     map[string]*flipt.Flag
+	segments  map[string]*flipt.Segment
+	rules     map[string]*flipt.Rule
+	evalRules map[string][]*storage.EvaluationRule
+}
+
+func newNamespace(key string) *namespace {
+	return &namespace{
+		resource:  &flipt.Namespace{Key: key, Name: strings.Title(key)},
+		flags:     map[string]*flipt.Flag{},
+		segments:  map[string]*flipt.Segment{},
+		rules:     map[string]*flipt.Rule{},
+		evalRules: map[string][]*storage.EvaluationRule{},
+	}
+}
 
 // snapshotFromFS constructs a storeSnapshot from the provided
 // fs.FS implementation.
 func snapshotFromFS(sources ...io.Reader) (*storeSnapshot, error) {
-	return nil, errors.New("not implemented")
+	s := storeSnapshot{
+		ns: map[string]*namespace{
+			defaultNs: {
+				resource: &flipt.Namespace{
+					Key:  defaultNs,
+					Name: "Default",
+				},
+			},
+		},
+		evalDists: map[string][]*storage.EvaluationDistribution{},
+	}
+
+	for _, reader := range sources {
+		doc := new(ext.Document)
+
+		if err := yaml.NewDecoder(reader).Decode(doc); err != nil {
+			return nil, err
+		}
+
+		if err := s.addDoc(doc); err != nil {
+			return nil, err
+		}
+
+	}
+	return &s, nil
+}
+
+func buildSnapshotHelper(logger *zap.Logger, source iofs.FS) ([]string, error) {
+	// This is the default variable + value for the FliptIndex. It will preserve its value if
+	// a .flipt.yml can not be read for whatever reason.
+	idx := FliptIndex{
+		Version: "1.0",
+		Inclusions: []string{
+			"**/features.yml", "**/features.yaml", "**/*.features.yml", "**/*.features.yaml",
+		},
+	}
+
+	// Read index file
+	inFile, err := source.Open(indexFile)
+	if err == nil {
+		if derr := yaml.NewDecoder(inFile).Decode(&idx); derr != nil {
+			return nil, derr
+		}
+	}
+
+	if err != nil && !errors.Is(err, iofs.ErrNotExist) {
+		return nil, err
+	} else {
+		logger.Debug("index file does not exist, defaulting...", zap.String("file", indexFile), zap.Error(err))
+	}
+
+	filenames := make([]string, 0)
+
+	for _, g := range idx.Inclusions {
+		f, err := iofs.Glob(source, g)
+		if err != nil {
+			logger.Error("malformed glob pattern for included files", zap.String("glob", g), zap.Error(err))
+			return nil, err
```

Comment:
Take it or leave it: The caller can log this out instead. We can just add context via wrap.
```suggestion
			return nil, fmt.Errorf("glob %q: %w", g, err)
```